client/cli/src/lib.rs

```diff
@@ -161,637 +144,227 @@ fn base_path(
+{
+	let app = T::clap();
 
-		// Note: while we would like the user to handle the exit themselves, we handle it here
-		// for backwards compatibility reasons.
-		let (exit_send, exit_recv) = std::sync::mpsc::channel();
-		let exit = exit.into_exit();
-		std::thread::spawn(move || {
-			block_on(exit);
-			let _ = exit_send.send(());
-		});
+	let mut full_version = sc_service::config::full_version_from_strs(
+		version.version,
+		version.commit
+	);
+	full_version.push_str("\n");
 
-		let mut export_fut = builder(config)?
-			.export_blocks(file, from.into(), to, json);
-		let fut = futures::future::poll_fn(|cx| {
-			if exit_recv.try_recv().is_ok() {
-				return Poll::Ready(Ok(()));
-			}
-			Pin::new(&mut export_fut).poll(cx)
-		});
+	let app = app
+		.name(version.executable_name)
+		.author(version.author)
+		.about(version.description)
+		.version(full_version.as_str());
 
-		let mut runtime = tokio::runtime::Runtime::new().unwrap();
-		runtime.block_on(fut)?;
-		Ok(())
-	}
-}
+	let matches = app.get_matches_from_safe(iter)?;
 
-/// Command ready to import the chain.
-pub struct ParseAndPrepareImport<'a> {
-	params: ImportBlocksCmd,
-	version: &'a VersionInfo,
+	Ok(T::from_clap(&matches))
 }
 
-impl<'a> ParseAndPrepareImport<'a> {
-	/// Runs the command and imports to the chain.
-	pub fn run_with_builder<C, G, E, F, B, S, Exit>(
-		self,
-		builder: F,
-		spec_factory: S,
-		exit: Exit,
-	) -> error::Result<()>
-	where S: FnOnce(&str) -> Result<Option<ChainSpec<G, E>>, String>,
-		F: FnOnce(Configuration<C, G, E>) -> Result<B, error::Error>,
-		B: ServiceBuilderCommand,
-		C: Default,
-		G: RuntimeGenesis,
-		E: ChainSpecExtension,
-		Exit: IntoExit
-	{
-		let mut config = create_config_with_db_path(
-			spec_factory,
-			&self.params.shared_params,
-			self.version,
-			None,
-		)?;
-		fill_import_params(
-			&mut config,
-			&self.params.import_params,
-			sc_service::Roles::FULL,
-			self.params.shared_params.dev,
-		)?;
-
-		let file: Box<dyn ReadPlusSeek + Send> = match self.params.input {
-			Some(filename) => Box::new(File::open(filename)?),
-			None => {
-				let mut buffer = Vec::new();
-				stdin().read_to_end(&mut buffer)?;
-				Box::new(Cursor::new(buffer))
-			},
-		};
-
-		// Note: while we would like the user to handle the exit themselves, we handle it here
-		// for backwards compatibility reasons.
-		let (exit_send, exit_recv) = std::sync::mpsc::channel();
-		let exit = exit.into_exit();
-		std::thread::spawn(move || {
-			block_on(exit);
-			let _ = exit_send.send(());
-		});
-
-		let mut import_fut = builder(config)?
-			.import_blocks(file, false);
-		let fut = futures::future::poll_fn(|cx| {
-			if exit_recv.try_recv().is_ok() {
-				return Poll::Ready(Ok(()));
-			}
-			Pin::new(&mut import_fut).poll(cx)
-		});
-
-		let mut runtime = tokio::runtime::Runtime::new().unwrap();
-		runtime.block_on(fut)?;
-		Ok(())
-	}
-}
+/// A helper function that initializes and runs the node
+pub fn run<F, G, E, FNL, FNF, SL, SF>(
```

Comment:
Why do we have a function that calls the `run` method of `RunCmd`. Not sure why we not call it directly.

Reply:
It might not be useful anymore indeed. Right now it's only helper to initialize & run the command. Should I delete?

Reply:
Yes, I think all of it could be moved into the `run` function of `RunCmd` itself.

Reply:
It's better to keep it for consistency with `run_subcommand`. The user can use `sc_cli::run` and `sc_cli::run_subcommand` depending on the arm in the match. It doesn't hurt to have it so I will keep it.

If you think it's a bad decision and it's best to delete it tell me and I will update the code.

Reply:
I think we wrote at the same time :grin: 

> Yes, I think all of it could be moved into the run function of RunCmd itself.

No it's not meant to be this way. Initialization must be separated from running the command so the user can change the whole logic with little effort. I might be wrong but here is the logic:

If the run command do the initialization for you than the user cannot insert something between the execution of the command and the initialization. The initialization does update the config.

Though I must say that none of our usecases are like that anymore (node, node-template, polkadot). Except maybe in cumulus where we will have to run 2 nodes but initialize only once.

(It also offers the user the opportunity to initialize without using the initializer but run the built-in command. This could be useful.)